addons/html_builder/static/src/core/builder_selection_restriction_plugin.js

```diff
@@ -0,0 +1,242 @@
+                anchorNode: anchorNode,
+                anchorOffset: anchorOffset,
+                focusNode: newAnchorNode,
+                focusOffset: newAnchorOffset,
+            });
+        }
+        this.correctSelectionOnUncrossable();
+        selection = this.dependencies.selection.getEditableSelection();
+        ({ focusNode, focusOffset } = selection);
+        this.dependencies.selection.setSelection({
+            anchorNode: focusNode,
+            anchorOffset: focusOffset,
+            focusNode: newFocusNode,
+            focusOffset: newFocusOffset,
+        });
+        this.correctSelectionOnUncrossable();
+    }
+
+    restrictSelectionInElement(selection, block) {
+        const { anchorNode, anchorOffset, focusNode, direction } = selection;
+        const isFocusInBlock = block.contains(focusNode);
+
+        if (!isFocusInBlock) {
+            let focusNode, focusOffset;
+            if (direction === DIRECTIONS.RIGHT) {
+                [focusNode, focusOffset] = getDeepestPosition(block, nodeSize(block));
+            } else {
+                [focusNode, focusOffset] = getDeepestPosition(block, 0);
+            }
+            this.dependencies.selection.setSelection({
+                anchorNode,
+                anchorOffset,
+                focusNode,
+                focusOffset,
+            });
+        }
+        this.correctSelectionOnUncrossable();
+    }
+
+    restrictSelectionInClosestDiv(ev) {
+        const { editableSelection, currentSelectionIsInEditable } =
+            this.dependencies.selection.getSelectionData();
+        if (!currentSelectionIsInEditable) {
+            return;
+        }
+        if (editableSelection.isCollapsed) {
+            return;
+        }
+
+        const closestSpecialBlock = closestElement(
+            editableSelection.anchorNode,
+            this.special_block_with_text_in_non_div.join(",")
+        );
+        const closestDiv = closestElement(editableSelection.anchorNode, "div");
+
+        if (closestSpecialBlock && !closestSpecialBlock.contains(closestDiv)) {
+            const closestParagraph = closestElement(editableSelection.anchorNode, "p");
+            this.restrictSelectionInElement(editableSelection, closestParagraph);
+        } else {
+            this.restrictSelectionInElement(editableSelection, closestDiv);
+        }
+    }
+
+    isNodeSelectionUncrossable(node, selectedNodes) {
+        return (
+            node.matches(this.uncrossable_element_selector) ||
+            selectedNodes.includes(closestElement(node, this.uncrossable_element_selector))
+        );
+    }
+
+    correctSelectionOnUncrossable() {
+        const selection = this.dependencies.selection.getEditableSelection();
+        const { anchorNode, anchorOffset, focusNode, direction } = selection;
+        const selectedNodes = this.dependencies.selection
+            .getTargetedNodes(selection)
+            .filter((node) => node.nodeType === Node.ELEMENT_NODE);
```

Comment:
You could directly use the `isElement` util function (from `dom_info.js`) :slightly_smiling_face: